geode-core/src/main/java/org/apache/geode/distributed/internal/MembershipManagerAdapter.java

```diff
@@ -0,0 +1,838 @@
+   * @throws InterruptedException if interrupted by another thread
+   * @throws TimeoutException if we wait too long for the member to go away
+   */
+  public boolean waitForDeparture(DistributedMember mbr, long timeoutMs)
+      throws TimeoutException, InterruptedException {
+    if (Thread.interrupted())
+      throw new InterruptedException();
+    boolean result = false;
+    // TODO - Move the bulk of this method to the adapter.
+    DirectChannel dc = directChannel;
+    InternalDistributedMember idm = (InternalDistributedMember) mbr;
+    long pauseTime = (timeoutMs < 4000) ? 100 : timeoutMs / 40;
+    boolean wait;
+    int numWaits = 0;
+    do {
+      wait = false;
+      if (dc != null) {
+        if (dc.hasReceiversFor(idm)) {
+          wait = true;
+        }
+        if (wait && logger.isDebugEnabled()) {
+          logger.info("waiting for receivers for {} to shut down", mbr);
+        }
+      }
+      if (!wait) {
+        wait = memberExists(idm);
+        if (wait && logger.isDebugEnabled()) {
+          logger.debug("waiting for {} to leave the membership view", mbr);
+        }
+      }
+      if (!wait) {
+        if (waitForSerialMessageProcessing(idm)) {
+          result = true;
+        }
+      }
+      if (wait) {
+        numWaits++;
+        if (numWaits > 40) {
+          throw new TimeoutException("waited too long for " + idm + " to be removed");
+        }
+        Thread.sleep(pauseTime);
+      }
+    } while (wait && (dc != null && dc.isOpen())
+        && !shutdownInProgress());
+    if (logger.isDebugEnabled()) {
+      logger.debug("operations for {} should all be in the cache at this point", mbr);
+    }
+    return result;
+  }
+
+  /**
+   * wait for serial executor messages from the given member to be processed
+   */
+  private boolean waitForSerialMessageProcessing(InternalDistributedMember idm)
+      throws InterruptedException {
+    // run a message through the member's serial execution queue to ensure that all of its
+    // current messages have been processed
+    boolean result = false;
+    OverflowQueueWithDMStats<Runnable> serialQueue =
+        clusterDistributionManager.getExecutors().getSerialQueue(idm);
+    if (serialQueue != null) {
+      final boolean done[] = new boolean[1];
+      final FlushingMessage msg = new FlushingMessage(done);
+      serialQueue.add(new SizeableRunnable(100) {
+        @Override
+        public void run() {
+          msg.invoke();
+        }
+
+        public String toString() {
+          return "Processing fake message";
+        }
+      });
+      synchronized (done) {
+        while (!done[0]) {
+          done.wait(10);
+        }
+        result = true;
+      }
+    }
+    return result;
+  }
+
+
+  /**
+   * Insert our own MessageReceiver between us and the direct channel, in order to correctly filter
+   * membership events.
+   *
+   *
+   */
+  class MyDCReceiver implements MessageListener {
```

Comment:
Would this be better as a `static` inner class? It acts solely through the `Membership` interface so it could take that kind of ref in the constructor. Also this could become a top-level class in that case.